### ISSUE TITLE
Find/replace dialog should remember its last position

### ISSUE THREAD

**Issue by ygoe** [notepad-plus-plus/notepad-plus-plus#905]
The dialog normally appears in the middle of the editor window, which is most likely in the way. So I move it a bit to the side to see the document while interacting with the search function. This position is retained when I close and re-open that window. But the position is reset again after Notepad++ was restarted. So this window placement should be included in the session state and be restored after quitting and restarting Notepad++.

This is pretty annoying when restarting Notepad++ often.

**Comment by xylographe** [comment]
Although I restart NPP only once a month on average, I would welcome this enhancement.

**Comment by dg9ngf** [comment]
I only noticed because I was starting it quite often today during bug hunting. Normally I'd keep it open until the next patchday as well.

**Comment by h-h-h-h** [comment]
> @xylographe:
I restart NPP only once a month on average

&nbsp;

> @dg9ngf:
Normally I'd keep it open until the next patchday as well.

And in between you constantly put Windows in hibernation, or not even that?

**Comment by xylographe** [comment]
No, Win7 remains active 24/7. Usually 3‒5 weeks before reboot is necessary because it becomes too slow.

**Comment by dg9ngf** [comment]
I can easily run my Windows 7 for several months, but Windows updates become too annyoing then so I give in and let them install. This definitely leads to a restart after more than a month.

**Comment by h-h-h-h** [comment]
What a luck that Notepad++ is green:
> Notepad++ is trying to reduce the world carbon dioxide emissions. When using less CPU power, the PC can throttle down and reduce power consumption, resulting in a greener environment.
(https://notepad-plus-plus.org/)

**Comment by MetaChuh** [comment]
we are closing this issue due to its age and multiple newer notepad++ versions since it was filed.

please update your notepad++ to the latest version [<kbd>Notepad++ 7.6.6</kbd>](https://notepad-plus-plus.org/download/v7.6.6.html).
if your issue persists, and it is still of relevance to you, please feel free to file a new issue.

thank you for your help and best regards.

-----
```
general notes:

please close your issues as soon as you are satisfied, to help
keeping the notepad++ issue tracker clean.

when creating a new issue, please provide all information as shown at the issue template.
your notepad++ debug info is part of the minimum requirements.
(? > Debug Info... > Copy debug info into clipboard).

please also conduct a search at the issue tracker beforehand, to check if a similar issue
already exists.

for general questions, or if it is not sure, whether your issue is directly related to
the notepad++ source code, please visit us at the notepad++ community forum, and
search if related topics exists. you are welcome to post either at similar
topics, or to create a new topic at Help Wanted or General Discussion.
```
[<kbd> &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; Click here to visit the Notepad++ Community Forum &emsp; &emsp; &emsp; &emsp; &emsp; &emsp;&emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; </kbd>](https://notepad-plus-plus.org/community/)

**Comment by sasumner** [comment]
A recent Community posting ( https://community.notepad-plus-plus.org/topic/18284/remember-find-replace-window-position ) has freshened this idea.  Let's take another look at it...

**Comment by sasumner** [comment]
A situation to stay out of would be restoring the Find window positioning to coordinates that no longer exist, resulting in a Find window that is potentially partially/fully offscreen, leaving the user mystified as to where it is, and likely no way to get it back to a viewable location.

Such a situation could occur by a change in monitor configuration between runs of Notepad++.

If it is detected that the Find window is (or would be) partially or fully off-screen at the remembered coordinates, then the current behavior (centering on the main Notepad++ window) should be preferred.

**Comment by Ekopalypse** [comment]
Another pitfall might be virtual desktops. As I don't have Win10 I can't test what the windows positioning APIs do return in such cases. It might look like an additional monitor, extended desktop or
just the main desktop.

**Comment by sasumner** [comment]
Looking at the code on how the _application_ window coordinates are saved, we see:

```
nppGUI._findWindowPos.right = posInfo.rcNormalPosition.right - posInfo.rcNormalPosition.left;;
nppGUI._findWindowPos.bottom = posInfo.rcNormalPosition.bottom - posInfo.rcNormalPosition.top;

```
It is sad that "right" and "bottom" are really "width" and "height". :-(

Should this "convention" be applied to new code for the find window positioning, for consistency's sake? I would say (gulp) probably Yes.

**Comment by Ekopalypse** [comment]
Or maybe this would be the time to introduce width and height? :-)

**Comment by xylographe** [comment]
> If it is detected that the Find window is (or would be) partially or fully off-screen at the remembered coordinates, then the current behavior (centering on the main Notepad++ window) should be preferred.

Unless the main Notepad++ window is partially or fully off-screen, thereby causing the Find window to be partially or fully off-screen, which must be avoided by all cost.

> Should this "convention" be applied to new code for the find window positioning,

I think not. This time get it right.

**Comment by sasumner** [comment]
I revise my original statement to be:

If it is detected that the Find window is (or would be) partially or fully off-screen at the remembered coordinates, then the current behavior (centering on the main Notepad++ window) should be preferred, _as long as this centering results in a Find window that is fully on-screen, otherwise we just let Windows position it as it should guarantee that it is fully viewable to the user._

**Comment by sasumner** [comment]
So #7898 is a PR for this issue, but it might be worth a little more discussion of that implementation.  I could do that in the PR comments section, but it feels better to put it here.

So, OK, the PR remembers and restores the Find window's position between runs of N++.  While N++ is running, Windows remembers the position, and when the user shuts down N++, the current position gets written to the `config.xml` file in the `FindWindowPos` attribute.  Easy.

The biggest part of the change deals with the case of position coordinates that can be outside the user's visible view.  A usage scenario for this is when a user is working on a laptop mounted on a docking station with the station providing access to one or more external monitors.  Say the user is running Notepad++ on one of these external monitors and has the Find window on a different external monitor (my common usage).  When the laptop is popped off the docking station (time to go to a Meeting!), Windows handles automatically moving the main N++ window to the monitor of the laptop, but the Find window is left at the coordinates it had when we were attached to the docking station.  

Before the PR, this situation would often cause a "lost" Find window, as its coordinates would be outside the coordinates of the laptop monitor (even reinvoking the Find command would not do anything to change the state of being "lost").  

After the PR, at the first (and, actually, every) invocation of the Find (or Replace or...) command, the location of the Find window's title bar is checked to see if it is viewable to the user.  If it is, then the Find window is made visible at its current location.  If not, the Find window will be moved to the center of the monitor its coordinates are closest to.  If the Find window is too wide to fit one monitor, it won't be centered but will be placed with its left edge along the left edge of the monitor.

**Comment by andrewc2** [comment]
I've noticed that this seems to work today however if you actuate a cntl-f to find while the window is already open, it resets the location to the center of the screen? Is this expected behavior? I swear it didn't used to be but not sure if I'm misremembering.

**Comment by alankilborn** [comment]
>I've noticed that this seems to work today

What "this" do you mean?

>if you actuate a cntl-f to find while the window is already open, it resets the location to the center of the screen

Yes, it is expected.

**Comment by andrewc2** [comment]
This as in, the find replace window remembers it's position after closing and reopening.

**Comment by alankilborn** [comment]
> This as in, the find replace window remembers it's position after closing and reopening.

And I presume you agree that is good behavior.

Above you can see I've opened an issue for the user manual to be updated with more info about pressing Ctrl+f twice in a row.

**Comment by andrewc2** [comment]
> And I presume you agree that is good behavior.

I do agree that's the correct behavior. 

However I don't prefer the find window resetting location upon repeated Ctrl+F, that would be nice as a toggleable option, but I understand that would be a feature request unrelated to this.
